README.md
# LifeSOS MQTT Client

[![PyPI version](https://badge.fury.io/py/lifesospy-mqtt.svg)](https://badge.fury.io/py/lifesospy-mqtt)

This application provides an MQTT client that interfaces with
[LifeSOS](http://lifesos.com.tw) alarm systems, sensors and switches.
It will publish the state of all devices to an MQTT broker, which can
then be consumed by any application that provides MQTT support, such
as Home Assistant or OpenHAB. It will also subscribe to topics on the
broker that allow the control of the alarm system (eg. arm, disarm)
and turn on/off any attached switches.

It was written for & tested with the LS-30 model, though it should also
work on the LS-10/LS-20 models. Note that in some markets, they may also
be labelled under the name of the distributor; eg. SecurePro in
Australia, WeBeHome in northern Europe.

## Requirements

- A network-connected LifeSOS LS-30, LS-20 or LS-10 alarm system.
- Something to run this application on. Assuming you want to leave it
running 24/7, a low-powered device like a Raspberry Pi or a NAS are ideal.
- [Python 3.5.3 or higher](https://www.python.org/downloads/) must be
installed.
- An MQTT broker to connect to; eg. [Mosquitto](https://mosquitto.org/)

## Installation

To install this application, run the following command:

    $ pip install lifesospy_mqtt

After installation of the application (and dependencies) has completed,
you should be able to start the application using the `lifesospy_mqtt`
command. For now, just try displaying the help using the `-h` switch
to check it is installed correctly. It should look something like this:

```
lifesospy_mqtt v0.10.1 - MQTT client to report state of LifeSOS security system and devices.

usage: lifesospy_mqtt [-h] [-e] [-v] [-d] [-w WORKDIR] [-c CONFIGFILE]
                      [-l [LOGFILE]] [-p [PIDFILE]]

optional arguments:
  -h, --help            show this help message and exit
  -e, --devices         list devices enrolled on base unit and exit
  -v, --verbose         display all logging output.
  -d, --daemon          put the application into the background after starting
  -w WORKDIR, --workdir WORKDIR
                        work directory used to store config, log and pid files
                        (default: ~/.lifesospy_mqtt)
  -c CONFIGFILE, --configfile CONFIGFILE
                        configuration file name (default: config.yaml)
  -l [LOGFILE], --logfile [LOGFILE]
                        if specified, will write to a daily rolling log file
                        (default: None)
  -p [PIDFILE], --pidfile [PIDFILE]
                        if specified, file will be create to record the
                        process ID and is used for locking (default: None)
```

## Configuration

When you run the application for the first time, it will create a new
configuration file and exit.

```
lifesospy_mqtt v0.10.1 - MQTT client to report state of LifeSOS security system and devices.

A default configuration file has been created:
~\.lifesospy_mqtt\config.yaml

Please edit any settings as needed then restart.
```

Open up the new config file. The following sections will step you
through filling in the settings.

#### LifeSOS settings

This section contains settings needed to interface with the LifeSOS
base unit / ethernet adapter.

```yaml
# Settings for the LifeSOS interface
lifesos:

  # Host and Port for the LifeSOS TCP Server to connect as a client,
  # or Port only to listen as a server for the LifeSOS TCP Client
  host: 192.168.1.100
  port: 1680

  # Master password, if needed by network interface
  password: ''
```

You will first need to determine whether your LifeSOS alarm system is
running in TCP SERVER or TCP CLIENT mode. Please refer to your manual
if unsure. This application is capable of supporting either mode,
though having LifeSOS in TCP SERVER MODE (ie. application will
run as a client and initiate connection) can make it easier to
diagnose issues, as connection failures will be logged.

If your LifeSOS ethernet adapter is configured in TCP SERVER mode:

- **host**: The host name / IP address of your LifeSOS ethernet adapter to connect to.
- **port**: The port number on your LifeSOS ethernet adapter.


If your LifeSOS ethernet adapter is configured in TCP CLIENT mode:

- **host**: *Unused*; this line must be deleted or prefixed with a hash (ie. `#host:`)
- **port**: The port number to listen for incoming connections from LifeSOS.

Remaining settings are for either mode:

- **password**: Master password, but **only if required by your LifeSOS
system**. My old LS-30 doesn't support it, but apparently it is a
configurable setting on newer models to require a password to issue
commands over the network interface.

#### MQTT settings

This section contains settings needed to connect to your MQTT broker.

```yaml
# Settings for the MQTT client
mqtt:

  # URI providing the details needed to connect to the MQTT broker
  # Refer https://github.com/mqtt/mqtt.github.io/wiki/URI-Scheme
  uri: mqtt://username:password@127.0.0.1:1883

  # Server certificate authority file/path/data; only for secured connection (mqtts/wss)
  #cafile: /some/certfile
  #capath: /some/folder
  #cadata:

  # Unique client identifier; no need to change unless running multiple instances
  client_id: lifesos
```

- **uri**: Provides the protocol, host name / IP address and port for
the MQTT broker, and a username/password for login (if required).
- **cafile**, **capath**, **cadata**: Optional settings used by HBMQTT
on secure connections; [reference](http://hbmqtt.readthedocs.io/en/latest/references/mqttclient.html#hbmqtt.client.MQTTClient.connect).
- **client_id**: Name to identify this application to the MQTT broker.

#### Translator settings

This section contains settings used by the component responsible for
translating between the MQTT client and the LifeSOS ethernet adapter.

```yaml
# Settings for the translator between LifeSOS and MQTT
translator:

  # To automatically configure devices in Home Assistant, ensure this line
  # matches the setting in Home Assistant's config file. Note any devices and
  # switches will need to be assigned a 'ha_name"' to be exported.
  # Refer https://www.home-assistant.io/docs/mqtt/discovery/
  ha_discovery_prefix: homeassistant

  # Topic and Payload to announce Home Assistant has come online. When received,
  # our MQTT client will send out the device configurations for discovery.
  ha_birth_topic: homeassistant/status
  ha_birth_payload: online

  # Provide a topic for the Base Unit here
  baseunit:
    topic: home/alarm
    ha_name: "LifeSOS"

  # List your enrolled devices here and provide a topic
  # Hint: Run with '-e' option to get a list of device ids
  devices:
    #- device_id: '012cba'
    #  topic: home/remote
    #- device_id: '123abc'
    #  topic: home/lounge/motion
    #  auto_reset_interval: 180
    #  ha_name: "Lounge Motion"

  # Uncomment any switches you own and provide a topic
  switches:
    #- switch_number: 1
    #  topic: home/lounge/heater
    #  ha_name: "Lounge Heater"
```

##### Base Unit settings

- **topic**: The topic name for the base unit (alarm panel).
Refer to the Topic section for more detail.

##### Device settings

Each enrolled device must have an entry here to link the unique
device identifier with a topic name. To determine the device identifier
for each device on your LifeSOS system, run `lifesospy_mqtt -e`.

```
lifesospy_mqtt v0.10.1 - MQTT client to report state of LifeSOS security system and devices.

Listing devices...
DeviceID '123456' for Controller zone 01-01, a RemoteController.
DeviceID '789abc' for Controller zone 01-02, a RemoteController.
DeviceID 'def123' for Burglar zone 01-01, a PIRSensor.
DeviceID '456789' for Burglar zone 01-02, a DoorMagnet.
DeviceID 'abcdef' for Burglar zone 03-01, a PIRSensor.
DeviceID 'fedcba' for Burglar zone 02-01, a PIRSensor.
DeviceID '987654' for Burglar zone 04-01, a PIRSensor.
DeviceID '321fed' for Burglar zone 04-02, a DoorMagnet.
DeviceID 'cba987' for Special zone 04-03, a TempSensor.
9 devices were found.
```

- **device_id**: Unique identifier for the device you want to publish
to the MQTT broker.
- **topic**: The topic name for the device. Refer to the
Topic section for more detail.
- **auto_reset_interval**: all trigger-based sensors (like the PIR
motion detector) only raise a single `Trigger` event when activated,
not an `On`/`Off` binary state often needed by home automation software.
This application will attempt to simulate a binary state by setting state
to `On` when triggered, then reset to `Off` after this duration (in seconds)
has elapsed. The default is 30 seconds when no value specified.

##### Switch settings

Each switch you own must have an entry here to link the switch number
with a topic name.

- **switch_number**: Number that identifies the switch on the base unit.
Should be a value between 1 and 16.
- **topic**: The topic name for the switch. Refer to the
Topic section for more detail.

##### Home Assistant settings

Settings prefixed with `ha_` are for [Home Assistant](https://www.home-assistant.io/)
support; specifically, it's [MQTT Discovery](https://www.home-assistant.io/docs/mqtt/discovery/)
feature. If you are not intending to use this application with Home
Assistant, these settings can either be ignored or removed.

- **ha_discovery_prefix**: the Home Assistant discovery prefix, as it is
listed in the Home Assistant configuration.yaml file.
- **ha_birth_topic**, **ha_birth_payload**: the topic and payload used
by Home Assistant to notify when it has come online
([refer here](https://www.home-assistant.io/docs/mqtt/birth_will/)). This
application will automatically send your LifeSOS configuration to Home
Assistant when it receives this message.
- **ha_name**: When this setting is listed under the device, it's
configuration will be exported to Home Assistant and the value will be
used as the display name for the device.

#### Logger settings

This section contains settings to configure application logging.

```yaml
# Settings to configure logging
# Valid severity levels are:
# critical, error, warning, info, debug
logger:

  default: info

  #namespaces:
  #  lifesos: debug
  #  hbmqtt: debug
```

- **default**: The default minimum severity level to log.
- **lifesos**: Minimum severity level for any log entries generated
directly from this application or the associated
[LifeSOSpy](https://github.com/rorr73/LifeSOSpy) library.
- **hbmqtt**: Minimum severity level for any log entries generated
from the [HBMQTT](https://github.com/beerfactory/hbmqtt) library.

## Topics

The topics published and subscribed to by this MQTT client are listed below.
In all cases, the parent topic name can be configured; the sub-topic names
and payloads are fixed to what is shown.

For maximum compatibility with other MQTT clients, all payloads are plain
text (UTF-8).

#### Base Unit Topics

**home/alarm** *Published*

Current state for the base unit. This is a superset of `operation_mode`; it
also takes into account the exit and entry delay. One of:

Value | Description
----- | -----------
`Disarm` | No burglar alarm will be triggered.
`Home` | Only perimeter sensors are armed; eg. outer doors, windows.
`Away` | All burglar sensors are armed.
`Monitor` | Disarmed, but triggered sensors will be logged; ie. for testing.
`AwayExitDelay` | Disarmed, but transitioning to Away mode after delay.
`AwayEntryDelay` | Burglar sensor triggered, delayed alarm pending.

**home/alarm/entry_delay** *Published*

When a burglar sensor is triggered, the base unit will wait this many seconds
before triggering the alarm (providing a brief window to disarm the system).

*Payload Example*: `15`

**home/alarm/exit_delay** *Published*

When the system is set to Away mode via a remote controller or keypad, it
will wait this many seconds before actually changing state (giving the owner
time to exit the premises).

*Payload Example*: `15`

**home/alarm/ha_state** *Published*

Provides the state of the base unit in a format recognised by Home Assistant's
`alarm_control_panel` platform. One of:

Value | Description
----- | -----------
`disarmed` | No burglar alarm will be triggered.
`armed_home` | Only perimeter sensors are armed; eg. outer doors, windows.
`armed_away` | All burglar sensors are armed.
`pending` | Pending mode change; ie. due to exit or entry delay.
`triggered` | Alarm has been triggered.

**home/alarm/is_connected** *Published*

Indicates if this LifeSOS MQTT client is operational and connected to the
base unit. One of: `True` or `False`.

**home/alarm/operation_mode** *Published*<br />**home/alarm/operation_mode/set** *Subscribed*

Gets or Sets the current operation mode. One of:

Value | Description
----- | -----------
`Disarm` | No burglar alarm will be triggered.
`Home` | Only perimeter sensors are armed; eg. outer doors, windows.
`Away` | All burglar sensors are armed.
`Monitor` | Disarmed, but triggered sensors will be logged; ie. for testing.

**home/alarm/rom_version** *Published*

Current ROM version reported by the Base Unit. May help in determining the
features available, and in diagnosing issues.

*Payload Example*: `02.4201/13/06`

**home/alarm/clear_status** *Subscribed*

When received, this application will clear the alarm/warning LEDs on base unit
and stop siren. No payload is required; it will be ignored.

**home/alarm/datetime/set** *Subscribed*

When received, the remote date/time will be set to the date/time specified in
the payload, or current local date/time if there is no payload.

This topic is useful when called periodically to fix drift and automatically
correct for Daylight Savings changes.

*Payload Example*: `2018-07-24-T13:57`

**home/alarm/event** *Published*

Details for an event when it occurs on the base unit. These are provided in
JSON format (if JSON is not available, refer `event_code` and `restore_code`
for an alternative).

Name | Description
---- | -----
`event_qualifier` | Context for the type of event. One of: `Event`, `Restore`, `Repeat`.
`event_category` | Category of event. One of: `Alarm`, `Supervisory`, `Trouble`, `OpenClose_Access`, `Bypass_Disable`, `Test_Misc`, `Automation`
`event_code` | Type of event. For a full list refer [here](https://github.com/rorr73/LifeSOSpy/blob/cae11d0d83190f873f2cdade9f8332797c558e9d/lifesospy/enums.py#L213).
`device_category` | Device category `code` and `description`, JSON encoded.
`zone` | Zone where the event occurred, if it originated from a device.
`user_id` | Unique identifier for user that generated the event, if any.

**home/alarm/event_code** *Published*<br />**home/alarm/restore_code** *Published*

Details for an event when it occurs on the base unit. Provided as an alternative
to the JSON version above, where JSON cannot be parsed by the MQTT client.

*Payload Example*: `ACPowerLoss`

#### Device Topics

**home/room/temp** *Published*

Current state for the device. The payload will depend on the type of device.

Type | Description
---- | -----------
Temperature | Temperature, in celcius.
Humidity | Humidity, as a percentage.
Light | Illuminance, as Lux.
AC Meter | Current, in amps.
Remote Control | Does not provide a state. Button press available through `event` topic only.
Keypad | Does not provide a state. Button press available through `event` topic only.
Magnet | One of: `Open`, `Closed`.
*Other* | Simulated only; *see below*. One of: `On`, `Off`.

Most sensors are trigger-based; eg. motion, window break, flood, gas, smoke, etc. These
do not normally provide a state; just a `Trigger` event when activated. Since most home
automation software requires a state, this application will attempt to simulate one by
providing `On` when triggered, and automatically reset to `Off` after a set duration.
By default, this is 30 seconds, although it can be customised via the configuration
setting `auto_reset_interval`.

Also note that motion sensors are designed to trigger as infrequently as possible, in
order to preserve battery life. After being triggered, they will not trigger again until
there has been no movement within the monitored area for at least 3 minutes. This can
make them unsuitable for home automation tasks, such as lighting an area when movement
is detected.

**home/room/temp/category/code** *Published*<br />**home/room/temp/category/description** *Published*

Code and Description for the category the device belongs to.

Code | Description
---- | -----------
`c` | `Controller` - eg. remote control, keypad.
`b` | `Burglar`- eg. magnet, motion, breakage sensors.
`f` | `Fire` - eg. smoke, gas, flood.
`m` | `Medical` - eg. medical wriststrap, emergency button.
`e` | `Special` - eg. temperature, humidity, light, power sensors.

**home/room/temp/device_id** *Published*

Unique identifier assigned to the device by the manufacturer.

*Payload Example*: `123abc`

**home/room/temp/rssi_db** *Published*

Wireless signal strength, in dB. Value in range between `0` to `100`.


docs/changelog.rst
Changelog
=========

0.11.6
------

- Updated python-daemon dependency to 2.2.4 to resolve issue with starting in daemon mode under Python 3.8 - `#15 <https://github.com/rorr73/LifeSOSpy_MQTT/issues/15>`__ 

0.11.5
------

- Fix PIR sensor not automatically resetting motion state when auto_reset_interval defined in configuration file - `#13 <https://github.com/rorr73/LifeSOSpy_MQTT/issues/13>`__

0.11.4
------

- Added low battery binary sensor for devices - `#9 <https://github.com/rorr73/LifeSOSpy_MQTT/issues/9>`__ (@bratanon)
- Fixed a deprecation warning that may be shown when loading the configuration file with PyYAML 5.1 or higher - `#11 <https://github.com/rorr73/LifeSOSpy_MQTT/issues/11>`__ (@bratanon)

0.11.3
------

- Perform graceful shutdown when SIGTERM signal received - `#7 <https://github.com/rorr73/LifeSOSpy_MQTT/issues/7>`__ (@bratanon)
- Fix issue where application doesn't shutdown immediately when signalled to exit - `#8 <https://github.com/rorr73/LifeSOSpy_MQTT/issues/8>`__

0.11.2
------

- Fix issue where Home Assistant will not reset from Triggered state when the alarm was triggered in Disarm mode (eg. due to tampering, panic button, environmental).

0.11.1
------

- Fix issue where devices shown as Unavailable in Home Assistant when MQTT connection is lost and subsequently reconnected.

0.11.0
------

- Changed the underlying MQTT client library from HBMQTT to Eclipse Paho, to reduce failures when reconnecting.
- Updated other dependent libraries.
- When a device is discovered that was not listed in the configuration file, log as warning and include device details.
- Support alternative Away/Disarm event pattern used by more recent LifeSOS units.

0.10.6
------

- Added auto config for Home Assistant to show device signal strength.

0.10.5
------

- Documentation moved out of README into `ReadTheDocs <http://lifesospy-mqtt.readthedocs.io>`__.

0.10.4
------

- Fix published HA auto config entry for alarm_control_panel, ahead of adding support to HA.


lifesospy_mqtt/config.py
"""
Configuration settings.
"""

import logging
import os
from typing import Any, Dict, Optional
from urllib.parse import urlparse, ParseResult
import yaml
from lifesospy.enums import SwitchNumber
from lifesospy_mqtt.const import SCHEME_MQTT, SCHEME_MQTTS
from lifesospy_mqtt.enums import LoggerLevel

_LOGGER = logging.getLogger(__name__)

CONF_AUTO_RESET_INTERVAL = 'auto_reset_interval'
CONF_BASEUNIT = 'baseunit'
CONF_CLIENT_ID = 'client_id'
CONF_DEFAULT = 'default'
CONF_DEVICE_ID = 'device_id'
CONF_DEVICES = 'devices'
CONF_HA_BIRTH_PAYLOAD = 'ha_birth_payload'
CONF_HA_BIRTH_TOPIC = 'ha_birth_topic'
CONF_HA_DISCOVERY_PREFIX = 'ha_discovery_prefix'
CONF_HA_NAME = 'ha_name'
CONF_HA_NAME_RSSI = 'ha_name_rssi'
CONF_HA_NAME_BATTERY = 'ha_name_battery'
CONF_HOST = 'host'
CONF_NAMESPACES = 'namespaces'
CONF_PASSWORD = 'password'
CONF_PORT = 'port'
CONF_SWITCH_NUMBER = 'switch_number'
CONF_SWITCHES = 'switches'
CONF_TOPIC = 'topic'
CONF_URI = 'uri'

GROUP_LIFESOS = 'lifesos'
GROUP_LOGGER = 'logger'
GROUP_MQTT = 'mqtt'
GROUP_TRANSLATOR = 'translator'

DEFAULT_LOGGERLEVEL = LoggerLevel.Info

DEFAULT_CONFIG = """
# Settings for the LifeSOS interface
""" + GROUP_LIFESOS + """:

  # Host and Port for the LifeSOS TCP Server to connect as a client,
  # or Port only to listen as a server for the LifeSOS TCP Client
  """ + CONF_HOST + """: 192.168.1.100
  """ + CONF_PORT + """: 1680
  
  # Master password, if needed by network interface
  """ + CONF_PASSWORD + """: ''

# Settings for the MQTT client
""" + GROUP_MQTT + """:

  # URI providing the details needed to connect to the MQTT broker
  # Refer https://github.com/mqtt/mqtt.github.io/wiki/URI-Scheme
  """ + CONF_URI + """: mqtt://username:password@127.0.0.1:1883
  
  # Unique client identifier; no need to change unless running multiple instances
  """ + CONF_CLIENT_ID + """: lifesos

# Settings for the translator between LifeSOS and MQTT
""" + GROUP_TRANSLATOR + """:

  # To automatically configure devices in Home Assistant, ensure this line
  # matches the setting in Home Assistant's config file. Note any devices and
  # switches will need to be assigned a '""" + CONF_HA_NAME + """"' to be exported.
  # Refer https://www.home-assistant.io/docs/mqtt/discovery/
  """ + CONF_HA_DISCOVERY_PREFIX + """: homeassistant
  
  # Topic and Payload to announce Home Assistant has come online. When received,
  # our MQTT client will send out the device configurations for discovery.
  """ + CONF_HA_BIRTH_TOPIC + """: homeassistant/status
  """ + CONF_HA_BIRTH_PAYLOAD + """: online

  # Provide a topic for the Base Unit here
  """ + CONF_BASEUNIT + """:
    """ + CONF_TOPIC + """: home/alarm
    """ + CONF_HA_NAME + """: "LifeSOS"
  
  # List your enrolled devices here and provide a topic
  # Hint: Run with '-e' option to get a list of device ids
  """ + CONF_DEVICES + """:
    #- """ + CONF_DEVICE_ID + """: '012cba'
    #  """ + CONF_TOPIC + """: home/remote
    #- """ + CONF_DEVICE_ID + """: '345def'
    #  """ + CONF_TOPIC + """: home/front/door
    #  """ + CONF_HA_NAME + """: "Front Door"
    #  """ + CONF_HA_NAME_RSSI + """: "Front Door RSSI"
    #  """ + CONF_HA_NAME_BATTERY + """: "Front Door Battery"
    #- """ + CONF_DEVICE_ID + """: '123abc'
    #  """ + CONF_TOPIC + """: home/lounge/motion
    #  """ + CONF_AUTO_RESET_INTERVAL + """: 180
    #  """ + CONF_HA_NAME + """: "Lounge Motion"
    #  """ + CONF_HA_NAME_RSSI + """: "Lounge Motion RSSI"
  
  # Uncomment any switches you own and provide a topic
  """ + CONF_SWITCHES + """:
    #- """ + CONF_SWITCH_NUMBER + """: 1
    #  """ + CONF_TOPIC + """: home/lounge/heater
    #  """ + CONF_HA_NAME + """: "Lounge Heater"
    #- """ + CONF_SWITCH_NUMBER + """: 2
    #  """ + CONF_TOPIC + """: home/room/switch02
    #- """ + CONF_SWITCH_NUMBER + """: 3
    #  """ + CONF_TOPIC + """: home/room/switch03
    #- """ + CONF_SWITCH_NUMBER + """: 4
    #  """ + CONF_TOPIC + """: home/room/switch04
    #- """ + CONF_SWITCH_NUMBER + """: 5
    #  """ + CONF_TOPIC + """: home/room/switch05
    #- """ + CONF_SWITCH_NUMBER + """: 6
    #  """ + CONF_TOPIC + """: home/room/switch06
    #- """ + CONF_SWITCH_NUMBER + """: 7
    #  """ + CONF_TOPIC + """: home/room/switch07
    #- """ + CONF_SWITCH_NUMBER + """: 8
    #  """ + CONF_TOPIC + """: home/room/switch08
    #- """ + CONF_SWITCH_NUMBER + """: 9
    #  """ + CONF_TOPIC + """: home/room/switch09
    #- """ + CONF_SWITCH_NUMBER + """: 10
    #  """ + CONF_TOPIC + """: home/room/switch10
    #- """ + CONF_SWITCH_NUMBER + """: 11
    #  """ + CONF_TOPIC + """: home/room/switch11
    #- """ + CONF_SWITCH_NUMBER + """: 12
    #  """ + CONF_TOPIC + """: home/room/switch12
    #- """ + CONF_SWITCH_NUMBER + """: 13
    #  """ + CONF_TOPIC + """: home/room/switch13
    #- """ + CONF_SWITCH_NUMBER + """: 14
    #  """ + CONF_TOPIC + """: home/room/switch14
    #- """ + CONF_SWITCH_NUMBER + """: 15
    #  """ + CONF_TOPIC + """: home/room/switch15
    #- """ + CONF_SWITCH_NUMBER + """: 16
    #  """ + CONF_TOPIC + """: home/room/switch16

# Settings to configure logging
# Valid severity levels are:
# critical, error, warning, info, debug
""" + GROUP_LOGGER + """:

  """ + CONF_DEFAULT + """: """ + str(DEFAULT_LOGGERLEVEL).lower() + """
  
  #""" + CONF_NAMESPACES + """:
  #  lifesospy: """ + str(LoggerLevel.Debug).lower() + """
  #  lifesospy_mqtt: """ + str(LoggerLevel.Debug).lower() + """
  #  paho.mqtt: """ + str(LoggerLevel.Debug).lower() + """
"""


class Config(object):
    """Contains the configuration settings."""

    def __init__(self, settings: Dict[str, Any], is_default: bool):
        self._lifesos = LifeSOSConfig(settings[GROUP_LIFESOS])
        self._mqtt = MQTTConfig(settings[GROUP_MQTT])
        self._translator = TranslatorConfig(settings[GROUP_TRANSLATOR])
        self._logger = LoggerConfig(settings.get(GROUP_LOGGER))
        self._is_default = is_default

    @property
    def is_default(self) -> bool:
        """True if default configuration file was created; otherwise, False."""
        return self._is_default

    @property
    def lifesos(self) -> 'LifeSOSConfig':
        """Configuration settings for the LifeSOS group."""
        return self._lifesos

    @property
    def logger(self) -> 'LoggerConfig':
        """Configuration settings for the Logger group."""
        return self._logger

    @property
    def mqtt(self) -> 'MQTTConfig':
        """Configuration settings for the MQTT group."""
        return self._mqtt

    @property
    def translator(self) -> 'TranslatorConfig':
        """Configuration settings for the Translator group."""
        return self._translator

    @classmethod
    def load(cls, config_path: str) -> Optional['Config']:
        """Load the configuration file, or create default if none exists."""
        is_default = False

        if os.path.isfile(config_path):
            # Specified file exists; we can simply use that
            _LOGGER.debug("Loading configuration file '%s'", config_path)
        else:
            # Create new configuration file with default settings
            _LOGGER.debug("Creating default configuration file '%s'",
                          config_path)
            try:
                with open(config_path, 'wt') as config_file:
                    config_file.write(DEFAULT_CONFIG)
            except Exception: # pylint: disable=broad-except
                _LOGGER.error("Failed to create default configuration file",
                              exc_info=True)
                return None
            is_default = True

        # Load the configuration settings
        try:
            with open(config_path, encoding='utf-8') as config_file:
                settings = yaml.load(config_file, Loader=yaml.SafeLoader) or {}
        except Exception: # pylint: disable=broad-except
            _LOGGER.error("Failed to parse configuration file", exc_info=True)
            return None

        # Return instance of the configuration settings
        return Config(settings, is_default)

    def __repr__(self):
        return "<{}: is_default={}, {}, {}, {}, {}>".format(
            self.__class__.__name__,
            self._is_default,
            self._lifesos,
            self._mqtt,
            self._translator,
            self._logger,
        )


class LifeSOSConfig(object):
    """Configuration settings for the LifeSOS interface."""

    def __init__(self, settings: Dict[str, Any]):
        self._host = settings.get(CONF_HOST)
        self._port = settings[CONF_PORT]
        self._password = settings.get(CONF_PASSWORD)

    @property
    def host(self) -> Optional[str]:
        """
        Host name or IP address for the LifeSOS Server if we are to be run as
        a client, or None if we are to run as a server.
        """
        return self._host

    @property
    def password(self) -> str:
        """Control password, if one has been assigned on the base unit."""
        return self._password

    @property
    def port(self) -> int:
        """
        Port number to connect to / listen on, depending on whether we're
        running as a client or server.
        """
        return self._port

    def __repr__(self):
        return "<{}: host={}, port={}, password={}>".format(
            self.__class__.__name__,
            self._host,
            self._port,
            "None" if not self._password else ''.ljust(len(self._password), '*'),
        )


class MQTTConfig(object):
    """Configuration settings for the MQTT client."""

    def __init__(self, settings: Dict[str, Any]):
        self._uri = urlparse(settings[CONF_URI])
        self._client_id = settings[CONF_CLIENT_ID]

        # Check URI specifies a supported scheme
        if not (self._uri.scheme == SCHEME_MQTT or self._uri.scheme == SCHEME_MQTTS):
            raise ValueError(
                "URI scheme '{}' is not supported".format(self._uri.scheme))

    @property
    def client_id(self) -> str:
        """Unique client identifier."""
        return self._client_id

    @property
    def uri(self) -> ParseResult:
        """URI providing the details needed to connect to the MQTT broker."""
        return self._uri

    def __repr__(self):
        return "<{}: uri={}, client_id={}>".format(
            self.__class__.__name__,
            self._uri,
            self._client_id,
        )


class TranslatorConfig(object):
    """Configuration settings for the translator between LifeSOS and MQTT."""

    def __init__(self, settings: Dict[str, Any]):
        self._ha_birth_payload = settings.get(CONF_HA_BIRTH_PAYLOAD)
        self._ha_birth_topic = settings.get(CONF_HA_BIRTH_TOPIC)
        self._ha_discovery_prefix = settings.get(CONF_HA_DISCOVERY_PREFIX)
        baseunit_settings = settings[CONF_BASEUNIT]
        self._baseunit = TranslatorBaseUnitConfig(baseunit_settings)

        self._devices = {}
        devices_settings = settings.get(CONF_DEVICES)
        if devices_settings:
            for device_settings in devices_settings:
                device_id = int(device_settings[CONF_DEVICE_ID], 16)
                self._devices[device_id] = \
                    TranslatorDeviceConfig(device_settings)

        self._switches = {}
        switches_settings = settings.get(CONF_SWITCHES)
        if switches_settings:
            for switch_settings in switches_settings:
                name = 'SW{:02d}'.format(switch_settings[CONF_SWITCH_NUMBER])
                switch_number = SwitchNumber.parse_name(name)
                if switch_number is None:
                    raise ValueError(
                        "{} is out of range".format(CONF_SWITCH_NUMBER))
                self._switches[switch_number] = \
                    TranslatorSwitchConfig(switch_settings)

    @property
    def baseunit(self) -> 'TranslatorBaseUnitConfig':
        """Configuration for the base unit."""
        return self._baseunit

    @property
    def devices(self) -> Dict[int, 'TranslatorDeviceConfig']:
        """Configuration for each enrolled device; lookup by device id."""
        return self._devices

    @property
    def ha_birth_payload(self) -> str:
        """Payload used to identify when Home Assistant comes online."""
        return self._ha_birth_payload

    @property
    def ha_birth_topic(self) -> str:
        """Topic used to identify when Home Assistant comes online."""
        return self._ha_birth_topic

    @property
    def ha_discovery_prefix(self) -> str:
        """Discovery prefix to auto configure devices in Home Assistant."""
        return self._ha_discovery_prefix

    @property
    def switches(self) -> Dict[SwitchNumber, 'TranslatorSwitchConfig']:
        """Configuration for each switch; lookup by switch number."""
        return self._switches

    def __repr__(self):
        return "<{}: baseunit={}, devices={}, switches={}>".format(
            self.__class__.__name__,
            self._baseunit,
            self._devices,
            self._switches,
        )


class TranslatorBaseUnitConfig(object):
    """Configuration settings for the translator specific to base unit."""

    def __init__(self, settings: Dict[str, Any]):
        self._topic = settings[CONF_TOPIC]
        self._ha_name = settings.get(CONF_HA_NAME)

    @property
    def ha_name(self) -> str:
        """Name to assign the base unit in Home Assistant."""
        return self._ha_name

    @property
    def topic(self) -> str:
        """Topic for the base unit."""
        return self._topic

    def __repr__(self):
        return "<{}: topic={}, ha_name={}>".format(
            self.__class__.__name__,
            self._topic,
            self._ha_name,
        )


class TranslatorDeviceConfig(object):
    """Configuration settings for the translator specific to a device."""

    def __init__(self, settings: Dict[str, Any]):
        self._topic = settings[CONF_TOPIC]
        self._auto_reset_interval = settings.get(CONF_AUTO_RESET_INTERVAL)
        self._ha_name = settings.get(CONF_HA_NAME)
        self._ha_name_rssi = settings.get(CONF_HA_NAME_RSSI)
        self._ha_name_battery = settings.get(CONF_HA_NAME_BATTERY)

    @property
    def auto_reset_interval(self) -> int:
        """Interval to wait before resetting state of a Trigger device."""
        return self._auto_reset_interval

    @property
    def ha_name(self) -> str:
        """Name to assign the device in Home Assistant."""
        return self._ha_name

    @property
    def ha_name_rssi(self) -> str:
        """Name to assign the device's signal strength in Home Assistant."""
        return self._ha_name_rssi

    @property
    def ha_name_battery(self) -> str:
        """Name to assign the device's battery status in Home Assistant."""
        return self._ha_name_battery

    @property
    def topic(self) -> str:
        """Topic for the device."""
        return self._topic

    def __repr__(self):
        return "<{}: topic={}, auto_reset_interval={}, ha_name={}, " \
               "ha_name_rssi={}, ha_name_battery={}>".format(
                self.__class__.__name__,
                self._topic,
                self._auto_reset_interval,
                self._ha_name,
                self._ha_name_rssi,
                self._ha_name_battery,
                )


class TranslatorSwitchConfig(object):
    """Configuration settings for the translator specific to a switch."""

    def __init__(self, settings: Dict[str, Any]):
        self._topic = settings[CONF_TOPIC]
        self._ha_name = settings.get(CONF_HA_NAME)

    @property
    def ha_name(self) -> str:
        """Name to assign the switch in Home Assistant."""
        return self._ha_name

    @property
    def topic(self) -> str:
        """Topic for the switch."""
        return self._topic

    def __repr__(self):
        return "<{}: topic={}, ha_name={}>".format(
            self.__class__.__name__,
            self._topic,
            self._ha_name,
        )


class LoggerConfig(object):
    """Configuration settings for logging."""

    def __init__(self, settings: Dict[str, Any]):
        self._default = DEFAULT_LOGGERLEVEL
        self._namespaces = {}

        if not settings:
            return

        self._default = LoggerLevel.parse_name(settings.get(CONF_DEFAULT))

        namespaces = settings.get(CONF_NAMESPACES)
        if namespaces:
            for namespace in namespaces.items():
                self._namespaces[namespace[0]] = \
                    LoggerLevel.parse_name(namespace[1])

    @property
    def default(self) -> LoggerLevel:
        """Default minimum severity level for logging."""
        return self._default

    @property
    def namespaces(self) -> Dict[str, LoggerLevel]:
        """Minimum severity level for a specific namespace."""
        return self._namespaces

    def __repr__(self):
        return "<{}: default={}, namespaces={}>".format(
            self.__class__.__name__,
            str(self._default),
            self._namespaces,
        )
